### PR TITLE
Add SAMA5D2 MCAN support

### PR DESCRIPTION
**Pull request by TimJTi**
## Summary
SAMA5D2 devices use (using Atmel/Microchip terminology) "MCAN" peripherals unlike the SAMAD3 devices which use "CAN" peripherals. This PR ensures the correct driver is used depending on the SAMA5 chip type selected.

Derived from the pre-existing samv7 arch MCAN driver.

## Impact
None expected.

## Testing
On custom board with SAMA5D27C-D1G processor and 2x MCAN interfaces. CAN TX and RX to FIFO buffers tested.

### CODE REVIEW

**Comment by acassis** [comment]
Please fix this small issue:
Warning: /home/runner/work/nuttx/nuttx/nuttx/arch/arm/src/sama5/hardware/_sama5d2x_memorymap.h:383:1: warning: #include outside of 'Included Files' section

**Comment by TimJTi** [comment]
There is a very specific comment in this file that the #include is in an unusual place, and it is quite common in many such file it seems?

File: arch/arm/src/sama5/sam_mcan.c [hunk 1]
```diff
@@ -0,0 +1,4202 @@
...
+            defined(CONFIG_SAMA5_MCAN1))
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Common definitions */
+
+#ifndef MIN
+#  define MIN(a,b) ((a < b) ? a : b)
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_mcan.c:68]:
```suggestion
#  define MIN(a,b) ((a) < (b) ? (a) : (b))
```

File: arch/arm/src/sama5/sam_mcan.c [hunk 2]
```diff
@@ -0,0 +1,4202 @@
...
+#endif
+
+#ifndef MAX
+#  define MAX(a,b) ((a > b) ? a : b)
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_mcan.c:72]:
```suggestion
#  define MAX(a,b) ((a) > (b) ? (a) : (b))
```

File: arch/arm/src/sama5/sam_mcan.c [hunk 3]
```diff
@@ -0,0 +1,4202 @@
...
+
+/* MCAN0 Configuration */
+
+#ifdef CONFIG_SAMA5_MCAN0
+
+/* Bit timing */
+
+#  define MCAN0_TSEG1  (CONFIG_SAMA5_MCAN0_PROPSEG + CONFIG_SAMA5_MCAN0_PHASESEG1 - 1)
+#  define MCAN0_TSEG2  (CONFIG_SAMA5_MCAN0_PHASESEG2 - 1)
+#  define MCAN0_BRP    ((uint32_t)(((float) SAMA5_MCANCLK_FREQUENCY / \
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_mcan.c:146]:
```suggestion
#  define MCAN0_BRP    ((uint32_t)(((float)SAMA5_MCANCLK_FREQUENCY / \
```

File: arch/arm/src/sama5/sam_mcan.c [hunk 4]
```diff
@@ -0,0 +1,4202 @@
...
+#  if MCAN0_TSEG2 > 128
+#    error Invalid MCAN0 TSEG2
+#  endif
+#  if MCAN0_SJW > 128
+#    error Invalid MCAN0 SJW
+#  endif
+
+#  define MCAN0_FTSEG1 (CONFIG_SAMA5_MCAN0_FPROPSEG + CONFIG_SAMA5_MCAN0_FPHASESEG1 - 1)
+#  define MCAN0_FTSEG2 (CONFIG_SAMA5_MCAN0_FPHASESEG2 - 1)
+#  define MCAN0_FBRP   ((uint32_t)(((float) SAMA5_MCANCLK_FREQUENCY / \
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_mcan.c:166]:
```suggestion
#  define MCAN0_FBRP   ((uint32_t)(((float)SAMA5_MCANCLK_FREQUENCY / \
```

File: arch/arm/src/sama5/sam_mcan.c [hunk 5]
```diff
@@ -0,0 +1,4202 @@
...
+#endif
+
+/* MCAN1 Configuration */
+
+#ifdef CONFIG_SAMA5_MCAN1
+  /* Bit timing */
+
+#  define MCAN1_TSEG1  (CONFIG_SAMA5_MCAN1_PROPSEG + CONFIG_SAMA5_MCAN1_PHASESEG1)
+#  define MCAN1_TSEG2  CONFIG_SAMA5_MCAN1_PHASESEG2
+#  define MCAN1_BRP    ((uint32_t)(((float) SAMA5_MCANCLK_FREQUENCY / \
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_mcan.c:436]:
```suggestion
#  define MCAN1_BRP    ((uint32_t)(((float)SAMA5_MCANCLK_FREQUENCY / \
```

File: arch/arm/src/sama5/sam_mcan.c [hunk 6]
```diff
@@ -0,0 +1,4202 @@
...
+#  if MCAN1_TSEG2 > 15
+#    error Invalid MCAN1 TSEG2
+#  endif
+#  if MCAN1_SJW > 15
+#    error Invalid MCAN1 SJW
+#  endif
+
+#  define MCAN1_FTSEG1 (CONFIG_SAMA5_MCAN1_FPROPSEG + CONFIG_SAMA5_MCAN1_FPHASESEG1)
+#  define MCAN1_FTSEG2 (CONFIG_SAMA5_MCAN1_FPHASESEG2)
+#  define MCAN1_FBRP   ((uint32_t)(((float) SAMA5_MCANCLK_FREQUENCY / \
```
Comment (pkarashchenko) [inline arch/arm/src/sama5/sam_mcan.c:453]:
```suggestion
#  define MCAN1_FBRP   ((uint32_t)(((float)SAMA5_MCANCLK_FREQUENCY / \
```

**Comment by acassis** [comment]
> There is a very specific comment in this file that the #include is in an unusual place, and it is quite common in many such file it seems?

Yes, many files was in the mainline before we added the CI nxstyle test. So, please use this opportunity to fix it, this way the file will be fixed and in the future when someone modify this file will not face this CI error again.

**Comment by TimJTi** [comment]
> > There is a very specific comment in this file that the #include is in an unusual place, and it is quite common in many such file it seems?
> 
> Yes, many files were in the mainline before we added the CI nxstyle test. So, please use this opportunity to fix it, this way the file will be fixed and in the future when someone modify this file will not face this CI error again.

But, from the comment in the file, it is in the middle of the file for a very specific reason:
```
/* SAMA5 Virtual (mapped) Memory Map
 *
 * board_memorymap.h contains special mappings that are needed when a ROM
 * memory map is used.  It is included in this odd location because it
 * depends on some the virtual address definitions provided above.
 */
```

So...moving the #include to the usual location does risk breaking things. For that reason I don't want to do anything to it with this PR. And, as said, the same comment is in multiple files so multiple arch files would need to change possibly breaking multiple builds.

Unless there's a very obvious fix, or this is not required (now?) for reasons I am not aware of,  would it be better to raise this as an issue and have a discussion about it then fix all such arch files?

**Comment by acassis** [comment]
> > > There is a very specific comment in this file that the #include is in an unusual place, and it is quite common in many such file it seems?
> > 
> > 
> > Yes, many files were in the mainline before we added the CI nxstyle test. So, please use this opportunity to fix it, this way the file will be fixed and in the future when someone modify this file will not face this CI error again.
> 
> But, from the comment in the file, it is in the middle of the file for a very specific reason:
> 
> ```
> /* SAMA5 Virtual (mapped) Memory Map
>  *
>  * board_memorymap.h contains special mappings that are needed when a ROM
>  * memory map is used.  It is included in this odd location because it
>  * depends on some the virtual address definitions provided above.
>  */
> ```
> 
> So...moving the #include to the usual location does risk breaking things. For that reason I don't want to do anything to it with this PR. And, as said, the same comment is in multiple files so multiple arch files would need to change possibly breaking multiple builds.
> 
> Unless there's a very obvious fix, or this is not required (now?) for reasons I am not aware of, would it be better to raise this as an issue and have a discussion about it then fix all such arch files?

Ah ok, I missed this comment, sorry!

**Pull request merged**